Shopify/cli
# Add config extension to app

### WHY are these changes introduced?
Add support for the new version app config specifications. Each section in the app config toml will be matched with one of the new `app_cofig` extension spec. A new extension instance will be created with the content of the section. All this new `app_config` extension instances will be uploaded with the `appDeploy` API request.
Creating a new `app_cofig` extension spec should be as simple as creating a new module inside the `@shopify/app` `extensions` folder. This should be an example of adding a new section to the app config schema
```ts
import {createConfigExtensionSpecification} from '../specification.js'
import {zod} from '@shopify/cli-kit/node/schema'

const PosEmbedmentSchema = zod.object({
  pos: zod
    .object({
      embedded: zod.boolean(),
    })
    .optional(),
})

const spec = createConfigExtensionSpecification({
  identifier: 'pos_embedment',
  schema: PosEmbedmentSchema,
})

export default spec

```

Co-authored-by: Alvaro Gutierrez [alvaro.gutierrezblanco@shopify.com](mailto:alvaro.gutierrezblanco@shopify.com)
Co-authored-by: Raluca Badoi [raluca.badoi@shopify.com](mailto:raluca.badoi@shopify.com)



### WHAT is this pull request doing?
- Creates the new `ConfigExtensionSpec`. It's a simplified version of the current `ExtensionSpec`. New management experience `app_config` remote specs will be mapped to this new `ConfigExtensionSpec` type.
- Modify `app loader` logic to support new app [config format ](https://docs.google.com/document/d/1EwN1mceHGdvepj8AxrgvchIlenzkiJs3tznfr_O1VSQ/edit). The new app schema is generated dynamically from the union of the subschema of each `app_config` extension spec
- Modify `app loader` logic to load each section of the app toml as a new `configExtensions` list inside the `app` object
- Deploy command will include the current `allExtensions` list and the new `configExtensions` list


### How to test your changes?
- Create a new spin instance `spin up partners -c shopify.branch=prototype-versioned-app-config -c partners.branch=add-app-config-management`
- Create a new `cli` app and run the `app config link` command to link it to a new remote app
- Edit the `shopify.app.toml` file and leave this content
```toml
name = "<YOUR_APP_NAME>"
version = "3"
client_id = "<YOUR_CLIENT_ID>"

[pos]
embedded = true
```
- Run the `app deploy` command to release a new version.
- Open the `partners` graphiql and run the following query and check that the `config` value is ok
```
 {
     app(apiKey: "1df452484825c0a1873948f2c4acdfb4") {
      extensionRegistrations {
        id
        uuid
        title
        type
        draftVersion {
          config
        }
        activeVersion {
          config
        }
      }
      configExtensionRegistrations {
        id
        uuid
        title
        type
        draftVersion {
          config
        }
        activeVersion {
          config
        }
      }
      dashboardManagedExtensionRegistrations {
        id
        uuid
        title
        type
        activeVersion {
          config
        }
        draftVersion {
          config
        }
      }
    }
  }
```
- Edit the `shopify.app.toml` file again and switch `embedded` to `false``
- Run the `deploy` command and the `graphql` query and the value should be updated


### Post-release steps



### Measuring impact

How do we know this change was effective? Please choose one:


### Checklist
